### DISCUSSION TITLE
Eddy Probe for N4Pro

### DISCUSSION THREAD

**Discussion by virajmalia** [OpenNeptune3D/OpenNept4une#244] [Ideas]
Hi all!
I was able to mount the BTT Eddy Duo probe on to the N4Pro extruder (I believe all N4s have the same extruder combo) using just 1 mounting screw, in the same spot as the factory probe.

I would like to share the config changes and the pictures here for other N4 variants to be tested, hopefully.

<details>
<summary> Pictures </summary>
    <img src="https://github.com/user-attachments/assets/66d638bb-816f-4ceb-a429-c3ba473aa612" width=50%><img src="https://github.com/user-attachments/assets/f7dbc59a-9c1a-480b-b832-6bda6a7815db" width=50%>
    <img src="https://github.com/user-attachments/assets/4480a030-5cfe-443b-ae32-9ef4b54261fb" width=50%><img src="https://github.com/user-attachments/assets/8adab5a3-034e-4c01-9fb7-f42ceef4e7cc" width=50%>
    <img src="https://github.com/user-attachments/assets/4efa22e2-5156-426b-83d9-73d92852002b" width=50%>
</details>

Video: https://youtube.com/shorts/bQfAcGh4rd0?feature=share

## Replace your existing `[probe]` and `[bed_mesh]` sections with the below configuration
```cfg
# The MCU section only applies to the Eddy USB. For Eddy Coil you will use the MCU name of the toolboard that you connected the Eddy Coil to.
[mcu eddy]
# Follow the steps at https://github.com/bigtreetech/Eddy to get the serial address.
serial: /dev/serial/by-id/usb-Klipper_rp2040_<> # This is the serial address of your eddy probe. This can be found by using the terminal of your klipper instance (typically through SSH) and using the command ```ls /dev/serial/by-id```
restart_method: command

[temperature_sensor btt_eddy_mcu]
sensor_type: temperature_mcu # Sets the type of sensor for Klipper to read
sensor_mcu: eddy # Sets the MCU of the eddy probe tempereature sensor
min_temp: 10 # Sets the minimum tempereature for eddys tempereature sensor to operate
max_temp: 100 # Sets the maximum tempereature for eddys tempereature sensor to operate

[probe_eddy_current btt_eddy]
sensor_type: ldc1612
z_offset: 2.5
i2c_mcu: eddy  # This value is good for the Eddy USB but would need to be adjusted for the Eddy Coil according to the MCU you have used.
i2c_bus: i2c0f # This value is good for the Eddy USB but would need to be adjusted for the Eddy Coil according to the I2C bus you have used.
# Measure the exact offsets below using the method described here: https://www.klipper3d.org/Probe_Calibrate.html#calibrating-probe-x-and-y-offsets
x_offset: -18.20
y_offset: 23.40

# This section is only relevant for Eddy USB. Comment it out for Eddy Coil.
[temperature_probe btt_eddy]
sensor_type: Generic 3950
sensor_pin: eddy:gpio26
horizontal_move_z: 2

[bed_mesh]
horizontal_move_z: 2
speed: 200
# For the mesh dimensions below, the coordinates need to be reachable by the center of the probe. To calculate coordinates that will work, use the formula below:
# mesh x min = position_min_x + greater_of (15mm or x_offset) <--- in this term, only consider the x offset if it is positive, ignore if negative.
# mesh y min = position_min_y + greater_of (15mm or y_offset) <--- in this term, only consider the y offset if it is positive, ignore if negative.
# mesh x max = position_max_x - greater_of (15mm or |x_offset|) <--- in this term, only consider the x offset if it is negative, ignore if positive.
# mesh y max = position_max_y - greater_of (15mm or |y_offset|) <--- in this term, only consider the y offset if it is negative, ignore if positive.
# Example: Consider that you have a 300 x 300 bed with the max x and y positions being 300 and the min being 0. Your probe offsets are -20 for X and 10 for Y
# For mesh x min we ignore the x offset term because it is negative. Therefore mesh x min = 15
# For mesh y min we do not ignore the y offset term because it is positive but it is less than 15 so we use 15. Therefore mesh y min = 15
# For mesh x max we do not ignore the x offset term because it is negative. It is also greater than 15. Therefore mesh x max = 280
# For mesh y max we ignore the y offset term because it is positive but it is less than 15 so we use 15. Therefore mesh y max = 285
# The final result would be mesh_min: 15, 15 mesh_max: 280, 285
mesh_min: 15,23.4  # modify these according to the above guide. If the probe cannot reach then you will get a klipper error when trying to scan a bed mesh.
mesh_max: 216.8,220 # modify these according to the above guide. If the probe cannot reach then you will get a klipper error when trying to scan a bed mesh.
probe_count: 9,9
algorithm: bicubic
#scan_overshoot: 4  #uncomment this section if you still have room left over on the X axis for some scan overshoot to product smoother movements and more accurate scanning.


###############################Macros and related################################
#This secton contains macros and related config sections. Some should be used, others are optional. Read the comment above each one to find out whether or not to uncomment it for your installation.

# Uncomment this if you are using Eddy as the probe AND the homing endstop AND would like to use the beta z-offset control
[save_variables]
filename: ~/printer_data/config/variables.cfg

# Uncomment this if you are using Eddy as the probe AND the homing endstop
[force_move]
enable_force_move: True # Allows a user to move the z axis down if they have no other means of homing Z and need to calibrate the Eddy.

# Uncomment this if you are using Eddy as the probe AND the homing endstop AND would like to use the beta z-offset control
[delayed_gcode RESTORE_PROBE_OFFSET]
initial_duration: 1.
gcode:
  {% set svv = printer.save_variables.variables %}
  {% if not printer["gcode_macro SET_GCODE_OFFSET"].restored %}
    SET_GCODE_VARIABLE MACRO=SET_GCODE_OFFSET VARIABLE=runtime_offset VALUE={ svv.nvm_offset|default(0) }
    SET_GCODE_VARIABLE MACRO=SET_GCODE_OFFSET VARIABLE=restored VALUE=True
  {% endif %}


# Uncomment this if you are using Eddy as the probe AND the homing endstop
# Take note of the lines that should only be used if you have a KNOMI installed.
[gcode_macro G28]
rename_existing: G28.1
gcode:
  G28.1 {rawparams}
  {% if not rawparams or (rawparams and 'Z' in rawparams) %}
    PROBE
    SET_Z_FROM_PROBE
  {% endif %}


# Uncomment this if you are using Eddy as the probe AND the homing endstop
[gcode_macro SET_Z_FROM_PROBE]
gcode:
    {% set cf = printer.configfile.settings %}
    SET_GCODE_OFFSET_ORIG Z={printer.probe.last_z_result - cf['probe_eddy_current btt_eddy'].z_offset + printer["gcode_macro SET_GCODE_OFFSET"].runtime_offset}
    G90
    G1 Z{cf.safe_z_home.z_hop}


# Uncomment this if you are using Eddy as the probe AND the homing endstop AND would like to use the beta z-offset control
[gcode_macro Z_OFFSET_APPLY_PROBE]
rename_existing: Z_OFFSET_APPLY_PROBE_ORIG
gcode:
  SAVE_VARIABLE VARIABLE=nvm_offset VALUE={ printer["gcode_macro SET_GCODE_OFFSET"].runtime_offset }


# Uncomment the lines in this macro if you are using Eddy as the probe AND the homing endstop AND would like to use the beta z-offset control
[gcode_macro SET_GCODE_OFFSET]
rename_existing: SET_GCODE_OFFSET_ORIG
variable_restored: False  # Mark whether the var has been restored from NVM
variable_runtime_offset: 0
gcode:
  {% if params.Z_ADJUST %}
    SET_GCODE_VARIABLE MACRO=SET_GCODE_OFFSET VARIABLE=runtime_offset VALUE={ printer["gcode_macro SET_GCODE_OFFSET"].runtime_offset + params.Z_ADJUST|float }
  {% endif %}
  {% if params.Z %}
    {% set paramList = rawparams.split() %}
    {% for i in range(paramList|length) %}
      {% if paramList[i]=="Z=0" %}
        {% set temp=paramList.pop(i) %}
        {% set temp="Z_ADJUST=" + (-printer["gcode_macro SET_GCODE_OFFSET"].runtime_offset)|string %}
        {% if paramList.append(temp) %}{% endif %}
      {% endif %}
    {% endfor %}
    {% set rawparams=paramList|join(' ') %}
    SET_GCODE_VARIABLE MACRO=SET_GCODE_OFFSET VARIABLE=runtime_offset VALUE=0
  {% endif %}
  SET_GCODE_OFFSET_ORIG { rawparams }

# This macro automates a lot of the frequency mapping process and simplifies the steps significantly.
[gcode_macro PROBE_EDDY_CURRENT_CALIBRATE_AUTO]
gcode:
  BED_MESH_CLEAR
  G28 X Y
  G90 # Abs positioning
  G1 X{ printer.toolhead.axis_maximum.x/2 } Y{ printer.toolhead.axis_maximum.y/2 } F6000
  {% if 'z' not in printer.toolhead.homed_axes %}
    SET_KINEMATIC_POSITION Z={ printer.toolhead.axis_maximum.z-1 } # Allows the user to work it down until it touches.
  {% endif %}
  PROBE_EDDY_CURRENT_CALIBRATE {rawparams}

#This macro is optional but useful if you want to run a rapid scan before each print. Simply uncomment it and add BED_MESH_SCAN to your print start code.
[gcode_macro BED_MESH_CALIBRATE]
rename_existing: BTT_BED_MESH_CALIBRATE
gcode:
  BTT_BED_MESH_CALIBRATE METHOD=rapid_scan
```

### Method 1 (only works for new sliced models)
Add `BED_MESH_CALIBRATE` command to your `Machine start gcode`.

### Method 2 (works for already uploaded models)
Replace your default `PRINT_START` gcode macro with:
```cfg
[gcode_macro PRINT_START]
{{ variable_small_print }}
{{ variable_outer_bed_force_activation_temp }}
{{ variable_force_outer_bed_heater }}

gcode:
    Frame_Light_ON
    Part_Light_ON
    G92 E0
    G90                       ; Use absolute coordinates
    BED_MESH_CLEAR

    {% set BED_TEMP = params.BED_TEMP|default(60)|float %}
    {% set BED_HEAT_SOAK_MINUTES = params.BED_HEAT_SOAK_MINUTES|default(0)|float %}
    {% set EXTRUDER_TEMP = params.EXTRUDER_TEMP|default(200)|float %}

    {{ start_bed_setup }}

    SET_BED_TEMPERATURE TARGET={BED_TEMP}                           ; Heat Bed to target temp
    BED_TEMPERATURE_WAIT MINIMUM={BED_TEMP-2} MAXIMUM={BED_TEMP+4}  ; Waits until the bed reaches close to target
    {% if BED_HEAT_SOAK_MINUTES > 0 %}
    RESPOND MSG="Waiting {BED_HEAT_SOAK_MINUTES} minutes for the bed to settle."
    G4 P{BED_HEAT_SOAK_MINUTES * 60000}
    {% endif %}

    CG28

    BED_MESH_CALIBRATE

    Smart_Park
    SET_FILAMENT_SENSOR SENSOR=filament_sensor ENABLE=1
    SET_HEATER_TEMPERATURE HEATER=extruder TARGET={EXTRUDER_TEMP}   ; Set and heat the final extruder temperature
    TEMPERATURE_WAIT SENSOR=extruder MINIMUM={EXTRUDER_TEMP-4} MAXIMUM={EXTRUDER_TEMP+10}  ; Wait for extruder to reach near target temperature
    LINE_PURGE                                                      ; KAMP Line Purge near print
    G92 E0                                                          ; Reset Extruder
    G1 Z2.0 F3000                                                   ; Move Z Axis up
    M117 Printing

```

## Contributions
I am new to making contributions to this project, so @halfmanbear please let me know if there is a way to enable support for eddy probe in the main branch of OpenNept4une repo itself. My changes are currently in https://github.com/virajmalia/OpenNept4une/tree/btt-eddy-config

Cheers and Happy New Year!

**Comment by thefjordkeeper** [comment]
Thanks for the code. I have a USB/CAN Eddy on the way and can't wait to try it.

**Reply by virajmalia** [reply]
Did you try?

**Comment by Kuyodie** [comment]
I'm trying to set up an Eddy probe, but I am struggling with the printer.cfg part. How do I know where to insert all of this information? I'm totally new to the coding part and I am struggling with it. I'm on a N4max if that helps. Does all of this get copied in the same section of the cfg or do I have to separate parts of it?

**Reply by zenmediasocial** [reply]
In general - you should put things like this in their own file - then use [include <newfile>.cfg] in the top of your printer.cfg 
This allows for much easier troubleshooting and keeps your printer.cfg organized. 
For this particular file - there are sections in the printer.cfg file you'd need to comment out because they are duplicates. I usually put a comment about why I'm commenting it out - so a year from now I remember.

**Reply by virajmalia** [reply]
>  I'm trying to set up an Eddy probe, but I am struggling with the printer.cfg part. How do I know where to insert all of this information? I'm totally new to the coding part and I am struggling with it. I'm on a N4max if that helps. Does all of this get copied in the same section of the cfg or do I have to separate parts of it?

You can replace the same sections in the printer.cfg. Keep a backup just in case. For n4max your measurements may be different in the bed mesh section.

**Comment by Openwrtfunboy** [comment]
Hello! Can i contact you in tg? Im ahving issues with n4p on openneptune with btt eddy usb. Nozzle going too low or too high from print to print. Ive done calibration, temperature calibration but i guess im missing the correct sequence. Also why you are not preheating your nozzle in print_start gcode?

**Reply by virajmalia** [reply]
After `Smart_Park`, there are a few lines that perform nozzle heating. I typically avoid preheating sooner in the start sequence because I tend to do a manual preheating sequence from the display, then change filament if required and then upload the model, so all this time allows for heating. You can use your own heating sequence. The only useful part in my `PRINT_START` macro is the `BED_MESH_CALIBRATE`. You can change everything else as per your requirements.

For nozzle too high or low means that btt-eddy cannot perform accurate readings. Make sure your btt-eddy is positioned correctly: https://github.com/bigtreetech/Eddy?tab=readme-ov-file#btt-eddy-dimensions-and-probe-location. Sorry, I don't use telegram.

**Comment by Kuyodie** [comment]
Unfortunately I stopped using my Neptune printers. I won’t be of much help I don’t think. I switched to Bambu printers 


Sent from Yahoo Mail for iPhone


On Saturday, July 5, 2025, 5:56 AM, Openwrtfunboy ***@***.***> wrote:




Hello! Can i contact you in tg? Im ahving issues with n4p on openneptune with btt eddy usb. Nozzle going too low or too high from print to print. Ive done calibration, temperature calibration but i guess im missing the correct sequence. Also why you are not preheating your nozzle in print_start gcode?

**Comment by DBSDBS** [comment]
Hi guys,

Can someone help me with nonstop error: !! Must calibrate probe_eddy_current first
I do exactly what is described [HERE](https://github.com/bigtreetech/Eddy?tab=readme-ov-file#2-drive-current-calibration) 
I receive this message:
<img width="568" height="102" alt="image" src="https://github.com/user-attachments/assets/234479cb-32ba-43eb-aa0f-6c53d5aa45a5" />
Then I click on the button: "Save config and restart" and after reconnection I try to do again G28 and receive again this  error: !! Must calibrate probe_eddy_current first

Then I checked the section [probe_eddy_current btt_eddy] in my  printer.cfg, saw that there is NO row:
reg_drive_current: 15

and put it in the section manually.... Again restart, again test G28 and again this  error: !! Must calibrate probe_eddy_current first

[probe_eddy_current btt_eddy]
sensor_type: ldc1612
z_offset: 2.5
i2c_mcu: eddy  # This value is good for the Eddy USB but would need to be adjusted for the Eddy Coil according to the MCU you have used.
i2c_bus: i2c0f # This value is good for the Eddy USB but would need to be adjusted for the Eddy Coil according to the I2C bus you have used.

reg_drive_current: 15
x_offset: 0
y_offset: 35

Can someone help me with this?

UPDATE:

Now I found at the end of the printer.cfg this section:

#*# <---------------------- SAVE_CONFIG ---------------------->
#*# DO NOT EDIT THIS BLOCK OR BELOW. The contents are auto-generated.
#*#
#*# [probe_eddy_current btt_eddy]
#*# reg_drive_current = 15

But the behavior of the printer is the same
